src/project.rs

```diff
@@ -31,68 +27,28 @@ pub enum ProjectLoadError {
     Io(io::Error),
 }
 
-impl From<FileLocationError> for ProjectLoadError {
-    fn from(err: FileLocationError) -> ProjectLoadError {
-        match err {
-            FileLocationError::NotFound => ProjectLoadError::ConfigNotFound,
-            FileLocationError::Io(io) => ProjectLoadError::Io(io),
+impl<'a, 'b> Project<'a, 'b> {
+    /// Given an absolute path to a shell.nix,
+    /// construct a Project and a ProjectConfig.
+    pub fn new(nix_file: &'a NixFile, gc_root: &'b Path) -> Project<'a, 'b> {
+        Project {
+            nix_file,
+            base_gc_root_path: gc_root,
         }
     }
-}
-
-impl Project {
-    /// Load a Project based on the current working directory,
-    /// locating a `shell.nix` configuration file in the current
-    /// directory.
-    pub fn from_cwd() -> Result<Project, ProjectLoadError> {
-        let shell_nix = locate_file::in_cwd("shell.nix")?;
-
-        Project::load(shell_nix, Project::default_gc_root_dir())
-    }
-
-    /// Default location in the user's XDG directories to keep
-    /// GC root pins
-    pub fn default_gc_root_dir() -> PathBuf {
-        let project_dir = ProjectDirs::from("com.github.target.lorri", "lorri", "lorri")
-            .expect("could not derive a gc root directory, please set XDG variables");
-
-        project_dir.cache_dir().to_path_buf()
-    }
-
-    /// Given a path to a shell.nix, construct a Project and a ProjectConfig.
-    pub fn load(shell_nix: PathBuf, gc_root: PathBuf) -> Result<Project, ProjectLoadError> {
-        let project_root = shell_nix
-            .parent()
-            // only None if `shell_nix` is "/"
-            .unwrap();
-
-        Ok(Project {
-            project_root: project_root.to_path_buf(),
-            shell_nix: shell_nix.clone(),
-            gc_root,
-        })
-    }
-
-    /// The project's human readable name
-    pub fn name(&self) -> &str {
-        &self
-            .project_root
-            .file_name()
-            .expect("unable to identify directory name of the project root")
-            .to_str()
-            .unwrap()
-    }
 
     /// Absolute path to the the project's primary entry points
     /// expression
-    pub fn expression(&self) -> PathBuf {
-        self.shell_nix.clone()
+    pub fn expression(&self) -> &NixFile {
+        &self.nix_file
     }
 
     /// Absolute path to the projects' gc root directory, for pinning
     /// build and evaluation products
     pub fn gc_root_path(&self) -> Result<PathBuf, std::io::Error> {
-        let path = self.gc_root.join(self.name()).join("gc_root");
+        // TODO: use a hash of the project’s abolute path here
+        // to avoid collisions
```

Comment:
is this a thing we're doing now?

Reply:
Nope, not yet. Haven’t got around to implementing that.